exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Query.kt

```diff
@@ -77,9 +87,43 @@ open class Query(override var set: FieldSet, where: Op<Boolean>?) : AbstractQuer
     }
 
     override fun withDistinct(value: Boolean): Query = apply {
+        if (value) {
+            require(distinctOn == null) { "DISTINCT cannot be used with the DISTINCT ON modifier. Only one of them should be applied." }
+        }
         distinct = value
     }
 
+    /**
+     * Specifies that the `SELECT` query should retrieve distinct results based on the given list of columns.
+     *
+     * This method can be used to set a `DISTINCT ON` clause for the query, which is supported by some SQL dialects
+     * (e.g., PostgreSQL, H2). The resulting query will retrieve rows that are distinct based on the specified columns.
+     *
+     * @param columns The columns to apply the `DISTINCT ON` clause.
+     * @return The current `Query` instance with the `DISTINCT ON` clause applied.
+     */
+    fun withDistinctOn(vararg columns: Column<*>): Query = apply {
+        require(!distinct) { "DISTINCT ON cannot be used with the DISTINCT modifier. Only one of them should be applied." }
+        distinctOn = (distinctOn ?: emptyList()) + columns
+    }
+
+    /**
+     * Specifies that the `SELECT` query should retrieve distinct results based on the given list of columns with sort orders.
+     * This method sets a `DISTINCT ON` clause and may reorder the results as indicated.
+     *
+     * This method can be used to set a `DISTINCT ON` clause for the query, which is supported by some SQL dialects
+     * (e.g., PostgreSQL, H2), along with an `ORDER BY` clause for the specified columns.
+     *
+     * @param columns The columns and their sort orders to apply the `DISTINCT ON` clause.
+     * @return The current `Query` instance with the `DISTINCT ON` clause and reordering applied.
+     */
+    fun withDistinctOn(vararg columns: Pair<Column<*>, SortOrder>): Query = apply {
+        require(!distinct) { "DISTINCT ON cannot be used with the DISTINCT modifier. Only one of them should be applied." }
+        @Suppress("SpreadOperator")
+        withDistinctOn(*columns.map { it.first }.toTypedArray())
+        return orderBy(*columns)
```

Comment:
note that if `withDistinctOn()` is called after `orderBy()` (which i think is technically possible) then the ordering may be incorrect (DISTINCT ON at least in postgresql requires the specified columns to appear in the ORDER BY expression in the same order), e.g.
```
table
  .selectAll()
  .orderBy(table.colThree)
  .withDistinctOn(table.colOne, table.colTwo)
```
 i think results in `ORDER BY table.col_three, table.col_one, table.col_two` which generates a SQL error.

a possible solution would be to manipulate the underlying `orderByExpressions` mutable list directly to prepend the list of columns passed in, instead of calling `orderBy()`. something like:
```
columns.forEachIndexed { index, column -> (orderByExpressions as MutableList).add(index, column) }
```
(since `private set` on the declaration `orderByExpressions` prohibits redefining the var)